Mathlib/Analysis/Complex/Periodic.lean

```diff
@@ -0,0 +1,216 @@
+    mul_div_cancel_left₀ _ (ofReal_ne_zero.mpr hh), exp_log hq]
+
+theorem qParam_left_inv_mod_period (hh : h ≠ 0) (z : ℂ) :
+    ∃ m : ℤ, invQParam h (𝕢 h z) = z + m * h := by
+  dsimp only [qParam, invQParam]
+  obtain ⟨m, hm⟩ := log_exp_exists (2 * ↑π * I * z / ↑h)
+  refine ⟨m, by rw [hm, mul_div_assoc, mul_comm (m : ℂ), ← mul_add, ← mul_assoc,
+    div_mul_cancel₀ _ two_pi_I_ne_zero, mul_add, mul_div_cancel₀ _ (mod_cast hh), mul_comm]⟩
+
+theorem abs_qParam_lt_iff (hh : 0 < h) (A : ℝ) (z : ℂ) :
+    abs (qParam h z) < Real.exp (-2 * π * A / h) ↔ A < im z := by
+  rw [abs_qParam, Real.exp_lt_exp, div_lt_div_right hh, mul_lt_mul_left_of_neg]
+  simpa using Real.pi_pos
+
+theorem qParam_tendsto (hh : 0 < h) : Tendsto (qParam h) I∞ (𝓝[≠] 0) := by
+  refine tendsto_nhdsWithin_of_tendsto_nhds_of_eventually_within _ ?_
+    (.of_forall fun q ↦ exp_ne_zero _)
+  rw [tendsto_zero_iff_norm_tendsto_zero]
+  simp only [norm_eq_abs, abs_qParam]
+  apply (tendsto_comap'_iff (m := fun y ↦ Real.exp (-2 * π * y / h)) (range_im ▸ univ_mem)).mpr
+  refine Real.tendsto_exp_atBot.comp (.atBot_div_const hh (tendsto_id.const_mul_atTop_of_neg ?_))
+  simpa using Real.pi_pos
+
+theorem invQParam_tendsto (hh : 0 < h) : Tendsto (invQParam h) (𝓝[≠] 0) I∞ := by
+  simp only [tendsto_comap_iff, comp_def, im_invQParam]
+  apply Tendsto.const_mul_atBot_of_neg (div_neg_of_neg_of_pos (neg_lt_zero.mpr hh) (by positivity))
+  exact Real.tendsto_log_nhdsWithin_zero_right.comp tendsto_norm_nhdsWithin_zero
+
+end qParam
+
+section PeriodicOnℂ
+
+variable (h : ℝ) (f : ℂ → ℂ)
+
+/-- The function `q ↦ f (invQParam h q)`, extended by a non-canonical choice of limit at 0. -/
+def cuspFunction : ℂ → ℂ :=
+  update (f ∘ invQParam h) 0 (limUnder (𝓝[≠] 0) (f ∘ invQParam h))
+
+theorem cuspFunction_eq_of_nonzero {q : ℂ} (hq : q ≠ 0) :
+    cuspFunction h f q = f (invQParam h q) :=
+  update_noteq hq ..
+
+theorem cuspFunction_zero_eq_limUnder_nhds_ne :
+    cuspFunction h f 0 = limUnder (𝓝[≠] 0) (cuspFunction h f) := by
+  conv_lhs => simp only [cuspFunction, update_same]
+  refine congr_arg lim (Filter.map_congr <| eventuallyEq_nhdsWithin_of_eqOn fun r hr ↦ ?_)
+  rw [cuspFunction, update_noteq hr]
+
+variable {f h}
+
+theorem eq_cuspFunction (hh : h ≠ 0) (hf : Periodic f h) (z : ℂ) :
+    (cuspFunction h f) (𝕢 h z) = f z := by
+  have : (cuspFunction h f) (𝕢 h z) = f (invQParam h (𝕢 h z)) := by
+    rw [cuspFunction, update_noteq, comp_apply]
+    exact exp_ne_zero _
+  obtain ⟨m, hm⟩ := qParam_left_inv_mod_period hh z
+  simpa only [this, hm] using hf.int_mul m z
+
+end PeriodicOnℂ
+
+section HoloOnC
+
+variable {h : ℝ} {f : ℂ → ℂ}
+
+/--
+Key technical lemma: the function `cuspFunction h f` is differentiable at the images of
+differentiability points of `f` (even if `invQParam` is not differentiable there).
+-/
+theorem differentiableAt_cuspFunction (hh : h ≠ 0) (hf : Periodic f h)
+    {z : ℂ} (hol_z : DifferentiableAt ℂ f z) :
+    DifferentiableAt ℂ (cuspFunction h f) (𝕢 h z) := by
```

Comment:
Do we want to also have a DifferentiableOn version? (given the section is called HoloOnC).

Reply:
I'm not sure. I can't really imagine settings in modular form theory where we would care about a `DifferentiableWithinAt` statement, or a `DifferentiableOn` for a set which isn't open. 

(Please don't read too much into the section name, that has survived unchanged from literally the first Lean code file I ever wrote!)

Reply:
There are also multiple possible formulations of a `DifferentiableOn` statement -- taking fwd / inverse image of a set under `qParam` or `invQParam` -- and it would be a pain to untangle which are true, and which of the true ones are useful. I suggest we wait, and see if any of them come up in practice as we work out more q-expansion theory.

Reply:
Sure I think thats sensible!